### DISCUSSION TITLE
2FA: FIDO / U2F / yubikey support

### DISCUSSION THREAD

**Discussion by DanielHeath** [rubygems/rubygems.org#2792] [Ideas]
The current two-factor auth implementation only supports OTP (google authenticator) codes. 

I'd like to also support physical keys that implement the FIDO standard (eg current-gen yubikeys), since that way I can recover my account if I lose my phone.

Has this come up anywhere else? The current design might need some tweaking as I'm unsure whether there is a nice way to use these on the command line.

**Comment by hsbt** [comment]
We welcomed your patch for this.

**Comment by DanielHeath** [comment]
The existing rails plugin for fido / u2f is GPL3 licenced; will that be 
compatible with this projects licencing requirements?

**Comment by bdewater** [comment]
Not sure which plugin you're referring to, but as an alternative https://github.com/cedarcode/webauthn-ruby is MIT licensed just like this project itself. 

For command line  https://github.com/sorah/clarion might be a solution, which uses the same gem.

**Comment by DanielHeath** [comment]
Webauthn is the previous spec; new keys use FIDO. 

There's https://github.com/castle/ruby-u2f (MIT licence) and https://github.com/TwoFactorAuth/ruby (AGPL3 / commercial licence). My gut feeling is that the latter is easier to integrate, though I could be wrong.

**Comment by bdewater** [comment]
> Webauthn is the previous spec; new keys use FIDO.

You're mixing a few things up. U2F was [added](https://security.googleblog.com/2014/10/strengthening-2-step-verification-with.html) to Chrome 38 in 2014, the FIDO Alliance [submitted](https://www.w3.org/Submission/2015/02/) FIDO 2.0 APIs to the W3C in 2015 for standardization, which happened last month when WebAuthn [became](https://www.w3.org/2019/03/pressrelease-webauthn-rec.html) a W3C standard. The spec specifically [mentions](https://www.w3.org/TR/webauthn/#conforming-authenticators-u2f) backwards compatibility with U2F.

FIDO 2.0 comprises of WebAuthn as the browser API and [Client To Authenticator Protocol](https://fidoalliance.org/specs/fido-v2.0-ps-20190130/fido-client-to-authenticator-protocol-v2.0-ps-20190130.html) for communicating with CTAP1 (U2F) and CTAP2 hardware.

**Comment by ecnelises** [comment]
> FIDO 2.0 comprises of WebAuthn as the browser API and Client To Authenticator Protocol for communicating with CTAP1 (U2F) and CTAP2 hardware.

So U2F will become legacy?

**Comment by bdewater** [comment]
Yes, but U2F remains supported. 

A big reason for the new API and CTAP2 hardware is that it besides being used as a second factor, it makes a secure multifactor passwordless login experience possible, by way of [resident keys](https://www.w3.org/TR/webauthn/#resident-credential) and testing for [user verification](https://www.w3.org/TR/webauthn/#user-verification) (instead of just user presence). E.g. the [Feitian Biopass FIDO2](https://www.ftsafe.com/products/FIDO2) would show something you have and something you are, but it could also be your [Android smartphone](https://venturebeat.com/2019/04/10/you-can-now-use-your-android-phone-as-a-2fa-security-key-for-google-accounts/).

**Comment by grzuy** [comment]
Prototype for this: https://github.com/rubygems/rubygems.org/pull/2108.

**Comment by simi** [comment]
This seems already implemented by https://github.com/rubygems/rubygems.org/pull/2865.
